Fix #14141: Split handler

SD-13 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #14141 .
2. This PR does the following: Splits Preferences Handler for better data validation.

## Essential Checklist

- [X] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [X] The linter/Karma presubmit checks have passed locally on your machine.
- [X] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [X] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

SD-13 committed:
some initial changes

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into split_handler

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into split_handler

SD-13 committed:
fix some backend tests

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into split_handler

SD-13 committed:
fix the failing backend test

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into split_handler

SD-13 committed:
Merge branch 'develop' into split_handler

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into split_handler

SD-13 committed:
add email preferences handler

SD-13 committed:
Merge branch 'split_handler' of https://github.com/SD-13/oppia into split_handler

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into split_handler

SD-13 committed:
Merge branch 'develop' into split_handler

SD-13 commented:
@vojtechjelinek please take a pass. Meanwhile, I will attempt to fix those tests and two more preferences page errors. Thanks

vojtechjelinek commented on core/controllers/profile.py:
```diff
@@ -148,24 +148,64 @@ def post(self, _):
+    @acl_decorators.can_manage_own_account
+    def put(self):
+        """Handles PUT requests."""
+        data = self.payload.get('data')
+        bulk_email_signup_message_should_be_shown = False
+
+        bulk_email_signup_message_should_be_shown = (
```
```suggestion
        data = self.payload.get('data')
        bulk_email_signup_message_should_be_shown = (
```

SD-13 replied:
done

vojtechjelinek commented on core/templates/services/user.service.ts:
```diff
@@ -74,7 +74,7 @@ export class UserService {
 
   async setProfileImageDataUrlAsync(
-      newProfileImageDataUrl: string): Promise<UpdatePreferencesResponse> {
+      newProfileImageDataUrl: string): Promise<ProfilePictureDataBackendDict> {
```
```suggestion
      newProfileImageDataUrl: string
  ): Promise<ProfilePictureDataBackendDict> {
```

SD-13 replied:
done

vojtechjelinek commented on core/templates/services/user-backend-api.service.ts:
```diff
@@ -136,19 +132,54 @@ export class UserBackendApiService {
+  }
+
+  async updateEmailPreferencesAsync(
+      data: boolean | EmailPreferencesBackendDict
```
Shouldn't this be only `EmailPreferencesBackendDict`

SD-13 replied:
yes, updated

vojtechjelinek commented on core/templates/services/user-backend-api.service.ts:
```diff
@@ -98,17 +94,17 @@ export class UserBackendApiService {
-      newProfileImageDataUrl: string): Promise<UpdatePreferencesResponse> {
-    return this.updatePreferencesDataAsync(
-      'profile_picture_data_url', newProfileImageDataUrl);
+  async setProfileImageDataUrlAsync(newProfileImageDataUrl: string):
+    Promise<ProfilePictureDataBackendDict> {
```
```suggestion
  async setProfileImageDataUrlAsync(
      newProfileImageDataUrl: string
  ): Promise<ProfilePictureDataBackendDict> {
```

SD-13 replied:
done

vojtechjelinek commented on core/templates/services/user-backend-api.service.ts:
```diff
@@ -34,29 +34,20 @@ export interface EmailPreferencesBackendDict {
   'subject_interests': string;
+}
+
+export interface ProfilePictureDataBackendDict {
```
Why is this together with `subscription_list`?

SD-13 replied:
Not for any specific reason. I am moving it to `PreferencesHandler`

SD-13 replied:
done

vojtechjelinek commented on core/templates/services/user-backend-api.service.ts:
```diff
@@ -136,19 +132,54 @@ export class UserBackendApiService {
-  async getPreferencesAsync(): Promise<PreferencesBackendDict> {
-    return this.http.get<PreferencesBackendDict>(this.PREFERENCES_DATA_URL)
-      .toPromise();
+  async getPreferencesAsync():
+    Promise<NonEmailPreferencesBackendDict> {
+    return this.http.get<NonEmailPreferencesBackendDict>(
+      this.PREFERENCES_DATA_URL).toPromise();
+  }
+
+  async getPreferencesProfilePictureDataUrlAsync():
+    Promise<ProfilePictureDataBackendDict> {
+    return this.http.get<ProfilePictureDataBackendDict>(
+      this.PREFERENCES_PROFILE_PICTURE_DATA_URL).toPromise();
+  }
+
+  async getEmailPreferencesAsync():
+    Promise<EmailPreferencesBackendDict> {
+    return this.http.get<EmailPreferencesBackendDict>(
```
```suggestion
  async getPreferencesAsync():
      Promise<NonEmailPreferencesBackendDict> {
    return this.http.get<NonEmailPreferencesBackendDict>(
      this.PREFERENCES_DATA_URL).toPromise();
  }

  async getPreferencesProfilePictureDataUrlAsync():
      Promise<ProfilePictureDataBackendDict> {
    return this.http.get<ProfilePictureDataBackendDict>(
      this.PREFERENCES_PROFILE_PICTURE_DATA_URL).toPromise();
  }

  async getEmailPreferencesAsync():
      Promise<EmailPreferencesBackendDict> {
    return this.http.get<EmailPreferencesBackendDict>(
```

SD-13 replied:
done

vojtechjelinek commented on core/templates/services/user-backend-api.service.ts:
```diff
@@ -136,19 +132,54 @@ export class UserBackendApiService {
+  ): Promise<NonEmailPreferencesBackendDict> {
+    return this.http.put<NonEmailPreferencesBackendDict>(
+      this.PREFERENCES_DATA_URL, {
+        update_type: updateType,
```
Why are we sending `update_type`?

SD-13 replied:
We currently use the `update_type` argument to know which value will be updated. But I am thinking about using the `data` argument as a dictionary to update values. For that `update_type` will not be needed anymore.

SD-13 replied:
Removed the use of `update_type`. Now we are sending all the data at once in a dictionary.

vojtechjelinek commented on core/templates/pages/preferences-page/preferences-page.component.ts:
```diff
@@ -96,18 +96,13 @@ export class PreferencesPageComponent {
     this.preventPageUnloadEventService.addListener();
     this.userBackendApiService.updatePreferencesDataAsync(
-      updateType, data).then((returnData) => {
+      updateType, data).then(() => {
```
```suggestion
      updateType, data
    ).then(() => {
```

SD-13 replied:
done

vojtechjelinek commented on core/templates/pages/preferences-page/preferences-page.component.ts:
```diff
@@ -136,6 +131,29 @@ export class PreferencesPageComponent {
+      data: EmailPreferencesBackendDict
+  ): void {
+    this.preventPageUnloadEventService.addListener();
+    this.userBackendApiService.updateEmailPreferencesAsync(
+      data).then((returnData) => {
```
```suggestion
    this.userBackendApiService.updateEmailPreferencesAsync(
      data
    ).then((returnData) => {
```

SD-13 replied:
done

vojtechjelinek commented on core/templates/pages/preferences-page/preferences-page.component.ts:
```diff
@@ -213,35 +224,55 @@ export class PreferencesPageComponent {
+      });
+    });
+
+
+    let emailPreferencesPromise = (
+      this.userBackendApiService.getEmailPreferencesAsync());
+
+    emailPreferencesPromise.then((data) => {
```
```suggestion
    let emailPreferencesPromise = (
      this.userBackendApiService.getEmailPreferencesAsync());
    emailPreferencesPromise.then((data) => {
```

SD-13 replied:
done

vojtechjelinek commented on core/templates/pages/preferences-page/preferences-page.component.ts:
```diff
@@ -213,35 +224,55 @@ export class PreferencesPageComponent {
 
     this.hasPageLoaded = false;
 
+    let preferencesProfilePictureDataUrlPromise = (
+      this.userBackendApiService.getPreferencesProfilePictureDataUrlAsync());
+
+    preferencesProfilePictureDataUrlPromise.then((data) => {
```
```suggestion
    let preferencesProfilePictureDataUrlPromise = (
      this.userBackendApiService.getPreferencesProfilePictureDataUrlAsync());
    preferencesProfilePictureDataUrlPromise.then((data) => {
```

SD-13 replied:
done

vojtechjelinek commented on core/handler_schema_constants.py:
```diff
@@ -54,6 +54,8 @@
     'PendingAccountDeletionPage',
     'PreferenceHandler',
     'PreferencesHandler',
+    'ProfilePictureDataUrlHandler',
+    'EmailPreferencesHandler',
```
Don't add these here rather immediately add the schemas

SD-13 replied:
okay, will do that.

SD-13 replied:
Added schema for those handler classes

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

SD-13 committed:
adds schema for few new handlers

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into split_handler

SD-13 commented:
@vojtechjelinek, I will be away for a few days. I will continue working on this after I return (probably on 8th October). Thanks,

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into split_handler

SD-13 committed:
fix few indentations

SD-13 committed:
move subscription list to preferences handler

SD-13 committed:
resolve merge conflict

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into split_handler

SD-13 committed:
adds schema and remove the use of update_type

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into split_handler

SD-13 commented:
@vojtechjelinek , Can you please provide an appropriate label for this PR. Thanks,

SD-13 committed:
fix backend test

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into split_handler

SD-13 committed:
Merge branch 'develop' into split_handler

SD-13 committed:
Merge branch 'develop' into split_handler

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into split_handler

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into split_handler

SD-13 committed:
Merge branch 'oppia:develop' into split_handler

SD-13 committed:
fix backend test coverage

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into split_handler

SD-13 commented:
Hey @vojtechjelinek, I took a quite long time to update this PR but finally I have addressed all your comments. The e2e tests are failing for some frontend errors which are appearing randomly, looks something like this

![Screenshot from 2023-01-30 00-01-16](https://user-images.githubusercontent.com/89520981/215857825-456054c0-a085-4a28-b08b-5d21a8cfdc99.png)

Please take another pass when you are free, in the meantime I will try to fix it. Thanks,

SD-13 commented:
And yes I know it really needs a design doc, I am working on it.

vojtechjelinek commented:
@SD-13 Not sure what is going on. Also, do you want a review?

SD-13 commented:
> @SD-13 Not sure what is going on.

Do you mean the errors I have mentioned? and yeah, please review this PR when you are free. Thanks,

vojtechjelinek commented:
@SD-13 Yep, not sure about the errors.

vojtechjelinek commented on core/controllers/profile.py:
```diff
@@ -258,26 +258,275 @@ def put(self) -> None:
+    HANDLER_ARGS_SCHEMAS = {
+        'GET': {},
+        'PUT': {
+            'data': {
```
```suggestion
            'profile_picture_data': {
```

SD-13 replied:
Okay, seems like we need to update it in the frontend part as well

vojtechjelinek replied:
yes

SD-13 replied:
I have added your suggested  changes but in frontend, as per the lint checks it's in camelCase

vojtechjelinek replied:
Makes sense.

vojtechjelinek commented on core/controllers/profile.py:
```diff
@@ -258,26 +258,275 @@ def put(self) -> None:
+    def put(self) -> None:
+        """Handles PUT requests."""
+        assert self.user_id is not None
+        assert self.normalized_payload is not None
```
Is this needed?

SD-13 replied:
I think so. We are inheriting `normalized_payload` from `BaseHandler` and [there](https://github.com/oppia/oppia/blob/ac68b404fef555b6982613effc483261365ea147/core/controllers/base.py#L220) it is set to `Optional`. But here we are not expecting `None` so, to cancel out that probability of having None I think we need this.

vojtechjelinek commented on core/controllers/profile.py:
```diff
@@ -318,47 +556,33 @@ def get(self) -> None:
-        elif update_type == 'preferred_language_codes':
-            user_services.update_preferred_language_codes(self.user_id, data)
-        elif update_type == 'preferred_site_language_code':
+        assert self.normalized_payload is not None
```
Is this needed?

SD-13 replied:
Same as above

vojtechjelinek reviewed (changes requested):
Took a partial pass.

SD-13 commented on core/controllers/profile.py:
```diff
@@ -258,26 +258,275 @@ def put(self) -> None:
+    HANDLER_ARGS_SCHEMAS = {
+        'GET': {},
+        'PUT': {
+            'data': {
```
@vojtechjelinek should we change `data` to `email_preferences_data` here?

vojtechjelinek replied:
Makes sense

SD-13 replied:
done. In frontend, I had to change it to camelCase.

SD-13 committed:
change variable names

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into split_handler

SD-13 commented:
@vojtechjelinek I have addressed all your comments and I need a help. Can you please help me to trace the reason of those frontend errors? I can help you to reproduce it. Also PTAL at rest of the changes. Thanks,

vojtechjelinek commented:
@SD-13 Please prepare a debugging doc for the errors, where you explain how to reproduce them and also what you tried to fix them.

vojtechjelinek commented on core/templates/pages/preferences-page/preferences-page.component.ts:
```diff
@@ -123,41 +132,68 @@ export class PreferencesPageComponent {
   savePreferredAudioLanguageCode(preferredAudioLanguageCode: string): void {
-    this._saveDataItem(
-      'preferred_audio_language_code', preferredAudioLanguageCode);
+    let data = this.PreferencesData();
+    data.preferred_audio_language_code = preferredAudioLanguageCode;
+    this._saveDataItem(data);
```
Can you simplify this somehow? I think you can still do saveItem(which, to what) and then retrieve the `PreferencesData` inside the `saveIteme`.

vojtechjelinek reviewed (changes requested):
Thanks! One comment. Also please provide proof video of your changes.

SD-13 committed:
Merge branch 'develop' into split_handler
